File: evaluer.mjs
#!/usr/bin/env node
// Atelier de la seance 5 : mesurer un prompt de classification.
//
// Vous ne modifiez QUE la constante CONSIGNE. Tout le reste est l'instrument de
// mesure : y toucher fausserait la comparaison.
//
//     node tp/05_prompt/evaluer.mjs
//
// Jumeau exact de evaluer.py : meme jeu de cas, meme score.

import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ICI = path.dirname(fileURLToPath(import.meta.url));
const BASE = (process.env.OLLAMA_BASE_URL || "http://localhost:11434").replace(/\/$/, "");
const MODELE = process.env.MODEL_BASE || "qwen2.5:3b";
const CAS = JSON.parse(readFileSync(path.join(ICI, "cas.json"), "utf8"));

// =====================================================================
// LA SEULE CHOSE QUE VOUS MODIFIEZ
// =====================================================================
const CONSIGNE = `Classe le ticket de support.`;
// =====================================================================

async function classer(texte) {
  const reponse = await fetch(`${BASE}/v1/chat/completions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model: MODELE,
      temperature: 0,
      max_tokens: 80,
      messages: [
        { role: "system", content: CONSIGNE },
        { role: "user", content: texte },
      ],
    }),
    signal: AbortSignal.timeout(60000),
  });
  const brut = (await reponse.json()).choices[0].message.content;
  const debut = brut.indexOf("{");
  const fin = brut.lastIndexOf("}");
  if (debut === -1 || fin === -1) return null;
  try {
    return JSON.parse(brut.slice(debut, fin + 1));
  } catch {
    return null;
  }
}

const total = CAS.cas.length;
let formes = 0, categories = 0, urgences = 0;
console.log(`modele : ${MODELE}   cas : ${total}\n`);

for (const [index, cas] of CAS.cas.entries()) {
  const numero = String(index + 1).padStart(2, " ");
  const obtenu = await classer(cas.texte);
  if (obtenu === null) {
    console.log(`  ${numero}. JSON illisible          <- ${cas.texte.slice(0, 44)}`);
    continue;
  }
  formes += 1;
  const bonneCategorie = obtenu.categorie === cas.categorie;
  const bonneUrgence = obtenu.urgence === cas.urgence;
  categories += bonneCategorie ? 1 : 0;
  urgences += bonneUrgence ? 1 : 0;
  const marque = bonneCategorie && bonneUrgence ? "ok " : "   ";
  console.log(`  ${numero}. ${marque} attendu ${cas.categorie}/${cas.urgence}`
    + `  obtenu ${obtenu.categorie}/${obtenu.urgence}`);
}

const pct = (n) => Math.floor((100 * n) / total);
console.log(`\n  JSON valide  : ${formes}/${total}  (${pct(formes)} %)`);
console.log(`  Categorie    : ${categories}/${total}  (${pct(categories)} %)`);
console.log(`  Urgence      : ${urgences}/${total}  (${pct(urgences)} %)`);
console.log("\nNotez ce score, modifiez CONSIGNE, relancez. Gardez la trace de "
  + "chaque version : elle est demandee au CC2.");

File: evaluer.py
#!/usr/bin/env python3
"""Atelier de la seance 5 : mesurer un prompt de classification.

Vous ne modifiez QUE la constante CONSIGNE. Tout le reste est l'instrument de
mesure : y toucher fausserait la comparaison.

    python3 tp/05_prompt/evaluer.py

Jumeau exact de evaluer.mjs : meme jeu de cas, meme score.
"""
import json
import os
import pathlib
import sys
import urllib.request

BASE = os.environ.get("OLLAMA_BASE_URL", "http://localhost:11434").rstrip("/")
MODELE = os.environ.get("MODEL_BASE", "qwen2.5:3b")
CAS = json.loads((pathlib.Path(__file__).parent / "cas.json").read_text(encoding="utf-8"))

# =====================================================================
# LA SEULE CHOSE QUE VOUS MODIFIEZ
# =====================================================================
CONSIGNE = """Classe le ticket de support."""
# =====================================================================


def classer(texte):
    charge = json.dumps({
        "model": MODELE,
        "temperature": 0,
        "max_tokens": 80,
        "messages": [{"role": "system", "content": CONSIGNE},
                     {"role": "user", "content": texte}],
    }).encode()
    requete = urllib.request.Request(
        f"{BASE}/v1/chat/completions", data=charge,
        headers={"Content-Type": "application/json"}, method="POST")
    with urllib.request.urlopen(requete, timeout=60) as reponse:
        brut = json.loads(reponse.read())["choices"][0]["message"]["content"]
    debut, fin = brut.find("{"), brut.rfind("}")
    if debut == -1 or fin == -1:
        return None
    try:
        return json.loads(brut[debut:fin + 1])
    except json.JSONDecodeError:
        return None


def main():
    total = len(CAS["cas"])
    formes, categories, urgences = 0, 0, 0
    print(f"modele : {MODELE}   cas : {total}\n")
    for i, cas in enumerate(CAS["cas"], 1):
        obtenu = classer(cas["texte"])
        if obtenu is None:
            print(f"  {i:2d}. JSON illisible          <- {cas['texte'][:44]}")
            continue
        formes += 1
        bonne_categorie = obtenu.get("categorie") == cas["categorie"]
        bonne_urgence = obtenu.get("urgence") == cas["urgence"]
        categories += bonne_categorie
        urgences += bonne_urgence
        marque = "ok " if bonne_categorie and bonne_urgence else "   "
        print(f"  {i:2d}. {marque} attendu {cas['categorie']}/{cas['urgence']}"
              f"  obtenu {obtenu.get('categorie')}/{obtenu.get('urgence')}")

    print(f"\n  JSON valide  : {formes}/{total}  ({100 * formes // total} %)")
    print(f"  Categorie    : {categories}/{total}  ({100 * categories // total} %)")
    print(f"  Urgence      : {urgences}/{total}  ({100 * urgences // total} %)")
    print("\nNotez ce score, modifiez CONSIGNE, relancez. Gardez la trace de "
          "chaque version : elle est demandee au CC2.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
